File: chazm-model/src/main/java/runtimemodels/chazm/model/relation/AchievesFactory.java
package runtimemodels.chazm.model.relation;

import runtimemodels.chazm.api.entity.Role;
import runtimemodels.chazm.api.entity.SpecificationGoal;
import runtimemodels.chazm.api.relation.Achieves;

/**
 * The {@linkplain AchievesFactory} interface defines the API for constructing {@linkplain Achieves} relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface AchievesFactory {

    /**
     * Constructs an {@linkplain Achieves}.
     *
     * @param role the {@linkplain Role} of the {@linkplain Achieves}.
     * @param goal the {@linkplain Achieves} of the {@linkplain Achieves}.
     * @return an {@linkplain Achieves}.
     */
    Achieves buildAchieves(Role role, SpecificationGoal goal);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/AchievesFactory.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.relation.Achieves

/**
 * The [AchievesFactory] interface defines the API for constructing [Achieves] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface AchievesFactory {
    /**
     * Constructs an [Achieves].
     *
     * @param role the [Role] of the [Achieves].
     * @param goal the [Achieves] of the [Achieves].
     * @return an [Achieves].
     */
    fun build(role: Role, goal: SpecificationGoal): Achieves
}
